src/Symfony/Component/HttpKernel/Controller/ContainerControllerResolver.php

```diff
@@ -82,10 +85,19 @@ protected function createController($controller)
      */
     protected function instantiateController($class)
     {
-        if ($this->container->has($class)) {
-            return $this->container->get($class);
+        try {
+            if ($this->container->has($class)) {
+                return $this->container->get($class);
+            }
+
+            return parent::instantiateController($class);
+        } catch (FatalThrowableError $e) {
```

Comment:
You need to be more specific than that. First of all: isn't it on the `$this->container->get(...)` that you want to catch the exception? Also, can you use the exception specific to this error, not a `Debug` exception?

Reply:
Actually, I'm not sure: this is Debug turning an `E_RECOVERABLE_ERROR` into an exception.
There is no way to catch this error without the debug component.
And any other exception might be unrelated in fact.
So this looks specific to me: improves DX by throwing to the user when applicable.

Reply:
Actually, I'm not sure: this is Debug turning an `E_RECOVERABLE_ERROR` into an exception.
There is no way to catch this error without the debug component.
And any other exception might be unrelated in fact.
So this looks specific to me: improves DX by throwing to the user when applicable.

Reply:
The problem I encountered into issue #25192 is that the Controller is not registered into the container, and the `parent::instantiateController()` method is called, which instantiate the controller without any argument.

So the try catch should only be for the `return parent::instantiateController($class);`.

Can we not use introspection in this method (`instantiateController`) to check that the constructor does not require arguments, and throw an exception here if it does?

Reply:
So it is maybe something to change on the `ControllerResolver` directly?

Reply:
I think I would go for that and throw an exception telling something similar to

"Your controller constructor requires argument but is neither declared as a public service nor tagged with `controller.service_arguments`"

But having instrospection is heavy work so I don't know if it worth it...

Reply:
the parent doens't know anything about the container, so here is the correct place, isn't it?
catching only the parent call looks sensible to me also

Reply:
@nicolas-grekas I think the point here is that `ControllerResolver::instantiateController()` cannot instantiate controllers requiring parameters in the constructor. So this method should check that the controller's constructor does not require any arguments and throw an exception if it does. This is not related to container.

If a controller is correctly configured as a public service or tagged with `controller.service_arguments`, then the controller will not be instantiate, and pulled from the container so the `ControllerResolver::instantiateController()` method will never be called, so no problem here...

Reply:
> So this method should check that the controller's constructor does not require any arguments and throw an exception if it does.

This supposes that the check is free, which is not the case. Here, we're in a performance critical code path. There is no better to do this check than the current one: try, and catch any error.